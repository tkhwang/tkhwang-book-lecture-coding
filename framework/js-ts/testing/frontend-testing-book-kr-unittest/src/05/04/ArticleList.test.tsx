import { render, screen, within } from "@testing-library/react";
import { ArticleList } from "./ArticleList";
import { items } from "./fixture";

/* 코드 5-13
test("목록을 표시한다", () => {
  render(<ArticleList items={items} />);
  const list = screen.getByRole("list");
  expect(list).toBeInTheDocument();
});
*/

test("제목을 표시한다", () => {
  render(<ArticleList items={items} />);
  expect(screen.getByRole("heading", { name: "기사 목록" })).toBeInTheDocument();
});

test("items의 수만큼 목록을 표시한다", () => {
  render(<ArticleList items={items} />);
  expect(screen.getAllByRole("listitem")).toHaveLength(3);
});

test("items의 수만큼 목록을 표시한다", () => {
  render(<ArticleList items={items} />);
  const list = screen.getByRole("list");
  expect(list).toBeInTheDocument();
  expect(within(list).getAllByRole("listitem")).toHaveLength(3);
});

test("목록에 표시할 데이터가 없으면 '게재된 기사가 없습니다'를 표시한다", () => {
  // 빈 배열을 items에 할당하여 목록에 표시할 데이터가 없는 상황을 재현한다.
  render(<ArticleList items={[]} />);
  // 존재하지 않을 것으로 예상하는 요소의 취득을 시도한다.
  const list = screen.queryByRole("list");
  // list가 존재하지 않는다.
  expect(list).not.toBeInTheDocument();
  // list가 null이다.
  expect(list).toBeNull();
  // '게재된 기사가 없습니다'가 표시됐는지 확인한다.
  expect(screen.getByText("게재된 기사가 없습니다")).toBeInTheDocument();
});

test("Snapshot: items의 수만큼 목록을 표시한다", () => {
  const { container } = render(<ArticleList items={items} />);
  expect(container).toMatchSnapshot();
});
