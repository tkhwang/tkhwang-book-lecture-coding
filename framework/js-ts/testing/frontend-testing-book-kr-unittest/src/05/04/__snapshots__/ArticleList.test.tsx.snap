// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot: items의 수만큼 목록을 표시한다 1`] = `
<div>
  <div>
    <h2>
      기사 목록
    </h2>
    <ul>
      <li>
        <h3>
          타입스크립트를 사용한 테스트 작성법
        </h3>
        <p>
          테스트 작성 시 타입스크립트를 사용하면 테스트의 유지 보수가 쉬워진다
        </p>
        <a
          href="/articles/howto-testing-with-typescript"
        >
          더 알아보기
        </a>
      </li>
      <li>
        <h3>
          Next.js의 링크 컴포넌트
        </h3>
        <p>
          Next.js는 화면을 이동할 때 링크 컴포넌트를 사용한다
        </p>
        <a
          href="/articles/nextjs-link-component"
        >
          더 알아보기
        </a>
      </li>
      <li>
        <h3>
          제스트로 시작하는 리액트 컴포넌트 테스트
        </h3>
        <p>
          제스트는 단위 테스트처럼 UI 컴포넌트를 테스트할 수 있다
        </p>
        <a
          href="/articles/react-component-testing-with-jest"
        >
          더 알아보기
        </a>
      </li>
    </ul>
  </div>
</div>
`;
